From d1a5008528c421036decc65e4c398cee6019d8f2 Mon Sep 17 00:00:00 2001
From: Dominic Farolino <dom@chromium.org>
Date: Thu, 10 Oct 2024 17:14:40 +0000
Subject: [PATCH] DOM: moveBefore() still runs the update selectedness algorithm

Per
https://docs.google.com/document/d/1qfYyvdK4zhzloABKeh0K1lHPm-SpnEcsWEE9UdDuoMk/edit, this CL adds tests for select/option/optgroup
move + selectedness semantics.

This CL also fixes a crash in ContainerNode::WillRemoveChildren(), which
is now reachable during an atomic move, via customizable select [1].
When an option element gets text-only rendered as a result of being
removed from a select element, it wipes *all* of its children, which
could include arbitrary content like iframes. The WillRemoveChildren()
path previously expected to never get called during an atomic move, but
this CL treats it the same as the WillRemoveChild() path, and gates the
child frame disconnector's use on non-SPAM flows.

[1]:
https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/html/forms/html_option_element.cc;l=581-583;drc=0afdaad03fc3fc271a751cb77dea8c8efb84b49f

R=masonf, nrosenthal

Bug: 40150299
Change-Id: I0e1064446b8a0359a62b1be152fec22fc9a583b9
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/5921538
Commit-Queue: Dominic Farolino <dom@chromium.org>
Reviewed-by: Noam Rosenthal <nrosenthal@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1366923}
---

--- a/third_party/blink/renderer/core/dom/container_node.cc
+++ b/third_party/blink/renderer/core/dom/container_node.cc
@@ -695,10 +695,14 @@
     DispatchChildRemovalEvents(child);
   }
 
-  CHECK(!GetDocument().StatePreservingAtomicMoveInProgress());
-  ChildFrameDisconnector(
-      *this, ChildFrameDisconnector::DisconnectReason::kDisconnectSelf)
-      .Disconnect(ChildFrameDisconnector::kDescendantsOnly);
+  // Only disconnect subframes in the non-state-preserving-atomic-move case,
+  // i.e., the traditional case where we intend to *fully* remove a node from
+  // the tree, instead of atomically re-inserting it.
+  if (!GetDocument().StatePreservingAtomicMoveInProgress()) {
+    ChildFrameDisconnector(
+        *this, ChildFrameDisconnector::DisconnectReason::kDisconnectSelf)
+        .Disconnect(ChildFrameDisconnector::kDescendantsOnly);
+  }
 }
 
 LayoutBox* ContainerNode::GetLayoutBoxForScrolling() const {
--- /dev/null
+++ b/third_party/blink/web_tests/external/wpt/dom/nodes/moveBefore/tentative/select-option-optgroup.html
@@ -0,0 +1,43 @@
+<!DOCTYPE html>
+<title>Selectedness is updated during moveBefore()</title>
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<body>
+
+<select>
+  <option value=A>A</option>
+  <optgroup label=Optgroup>
+    <option value=B>B</option>
+  </optgroup>
+  <option value=C>C</option>
+</select>
+
+<script>
+test(t => {
+  const optionA = document.querySelector('option[value=A]');
+  const optionB = document.querySelector('option[value=B]');
+  const optionC = document.querySelector('option[value=C]');
+
+  assert_true(optionA.selected, "A is selected by default");
+  assert_false(optionB.selected, "B not selected before first move");
+  assert_false(optionC.selected, "C not selected before first move");
+
+  // `moveBefore()`, like `insertBefore()` and kin, leaves a text-only-rendered
+  // non-select-owned <option> tag in the "selected" state.
+  document.body.moveBefore(optionA, null);
+  assert_true(optionA.selected, "A STILL selected after it is moved out");
+  assert_true(optionB.selected, "B becomes selected after first move");
+  assert_false(optionC.selected, "C is not selected after first move");
+
+  document.body.moveBefore(optionB, null);
+  assert_true(optionA.selected, "A STILL selected after second move");
+  assert_true(optionB.selected, "B STILL selected after it is moved out");
+  assert_true(optionC.selected, "C becomes selected after second move");
+
+  // Move A back into <select> before C.
+  document.querySelector('select').moveBefore(optionA, optionC);
+  assert_true(optionA.selected, "A STILL selected after it is moved back in");
+  assert_true(optionB.selected, "B STILL selected after third move");
+  assert_false(optionC.selected, "C no longer selected after third move");
+}, "Option selectedness is updated on option and optgroup DOM moves");
+</script>
